actions/drain_node_handler.go

```diff
@@ -107,74 +94,71 @@ func (h *drainNodeHandler) Handle(ctx context.Context, data interface{}) error {
-	for _, pod := range pods {
-		pod := pod
-
-		g.Go(func() error {
-			err := h.deletePod(ctx, pod)
-			if err != nil {
-				return err
-			}
-			return h.waitPodTerminated(ctx, log, pod)
-		})
+func (h *drainNodeHandler) taintNode(ctx context.Context, node *v1.Node) error {
+	if node.Spec.Unschedulable {
+		return nil
 	}
 
-	if err := g.Wait(); err != nil {
-		return fmt.Errorf("deleting pods: %w", err)
+	err := patchNode(ctx, h.clientset, node, func(n *v1.Node) error {
+		n.Spec.Unschedulable = true
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("patching node unschedulable: %w", err)
 	}
-
 	return nil
 }
 
-func (h *drainNodeHandler) deletePod(ctx context.Context, pod v1.Pod) error {
-	b := backoff.WithContext(backoff.WithMaxRetries(backoff.NewConstantBackOff(h.cfg.podDeleteRetryDelay), h.cfg.podDeleteRetries), ctx) // nolint:gomnd
-	action := func() error {
-		err := h.clientset.CoreV1().Pods(pod.Namespace).Delete(ctx, pod.Name, metav1.DeleteOptions{})
-		if err != nil {
-			// Pod is not found - ignore.
-			if apierrors.IsNotFound(err) {
-				return nil
-			}
+func (h *drainNodeHandler) evictNodePods(ctx context.Context, log logrus.FieldLogger, node *v1.Node) error {
+	pods, err := h.listNodePodsToEvict(ctx, node)
+	if err != nil {
+		return err
+	}
 
-			// Pod is misconfigured - stop retry.
-			if apierrors.IsInternalError(err) {
-				return backoff.Permanent(err)
-			}
-		}
+	log.Infof("evicting %d pods", len(pods))
 
-		// Other errors - retry.
+	if err := h.sendPodsRequests(ctx, pods, h.evictPod); err != nil {
+		return fmt.Errorf("sending evict pods requests: %w", err)
+	}
+
+	return h.waitNodePodsTerminated(ctx, node)
+}
+
+func (h *drainNodeHandler) deleteNodePods(ctx context.Context, log logrus.FieldLogger, node *v1.Node) error {
+	pods, err := h.listNodePodsToEvict(ctx, node)
+	if err != nil {
 		return err
 	}
-	if err := backoff.Retry(action, b); err != nil {
-		return fmt.Errorf("deleting pod %s in namespace %s: %w", pod.Name, pod.Namespace, err)
+
+	log.Infof("forcefully deleting %d pods", len(pods))
+
+	if err := h.sendPodsRequests(ctx, pods, h.deletePod); err != nil {
+		return fmt.Errorf("sending delete pods requests: %w", err)
 	}
-	return nil
+
+	return h.waitNodePodsTerminated(ctx, node)
 }
 
-// taintNode to make it unshedulable.
-func (h *drainNodeHandler) taintNode(ctx context.Context, node *v1.Node) error {
-	if node.Spec.Unschedulable {
-		return nil
-	}
+func (h *drainNodeHandler) sendPodsRequests(ctx context.Context, pods []v1.Pod, f func(context.Context, v1.Pod) error) error {
+	const batchSize = 5
 
-	err := patchNode(ctx, h.clientset, node, func(n *v1.Node) error {
-		n.Spec.Unschedulable = true
-		return nil
-	})
-	if err != nil {
-		return fmt.Errorf("patching node unschedulable: %w", err)
+	for _, batch := range lo.Chunk(pods, batchSize) {
+		g, ctx := errgroup.WithContext(ctx)
+		for _, pod := range batch {
+			pod := pod
+			g.Go(func() error { return f(ctx, pod) })
```

Comment:
Before with single request for pod we were retrying request with backoff. Would be good to wrap f(ctx, pod) into retry here to.

Reply:
Requests are wrapped inside the functions with different params for different call. Check EvictPod, DeletePod methods.

Reply:
Ah, ok, cool!